### PR TITLE
Check for the correct pairs of (transformers, masking scheme)

### LINKED ISSUE #237: Test transformers architectures and link them to the correct masking tasks.
- The issue is related to documenting and handling transformers dependencies. HF library is not  a modular toolbox of building blocks and does not separate the transformer architecture from the masking schema used in the original published work. Besides, HF transformers  library supports different architectures related to various tasks: Sequence Classification,  Language Modelling, Named Entity Recognition, Question/Answering, Summarization and Translation.  For session-based and sequential-based recommendation, we only need the architectures that support the three first tasks. 

- [x] We need to list those architecture and add a table in our Core API doc that shows the user the architectures he could use from NLP domain. 

- [x]  We also need to link each architecture to the possible pre-registered masking schemes: add a column in the doc table + raise an exception in the code when the wrong masking task is selected. P.s: for custom masking schema implemented by the user, we don't perform this check.

### PR DESCRIPTION
**Pull request by sararb**
1. Please write a description in this text box of the changes that are being
   made.

This PR adds documentation about the supported HF transformers architectures. It also implements a check in the code to raise an exception when the Transformer architecture is used with the wrong masking scheme.  Note that we only check for default masking classes. If the user defines and uses a custom masking class, this check is ignored. 

2. Please ensure that you have written units tests for the changes made/features
   added.

A unit test is defined: it checks if a Value Error is raised when the wrong pair (transformers, masking) is passed to TransformerBlock. 

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

closes #237

### CODE REVIEW

File: docs/source/resources.md [hunk 1]
```diff
@@ -14,6 +14,44 @@ Transformers4Rec is part of the NVIDIA Merlin ecosystem for Recommender Systems.
...
 - [HugeCTR](https://github.com/NVIDIA/HugeCTR) - A GPU-accelerated recommender framework designed to distribute training across multiple GPUs and nodes and estimate Click-Through Rates (CTRs). 
 
+
+## Supported HuggingFace architectures and pre-training approaches
+
+Transformers4Rec supports the four following [masking tasks](https://nvidia-merlin.github.io/Transformers4Rec/main/model_definition.html#sequence-masking): 
+
+|Accronym| Definition|
+|---------|--------------|
+| CLM    | Causal Language Modelling|
```
Comment (gabrielspmoreira) [inline docs/source/resources.md:24]:
```suggestion
| CLM    | Causal Language Modeling|
```

File: docs/source/resources.md [hunk 2]
```diff
@@ -14,6 +14,44 @@ Transformers4Rec is part of the NVIDIA Merlin ecosystem for Recommender Systems.
...
+| MLM    | Masked Language Modelling|
```
Comment (gabrielspmoreira) [inline docs/source/resources.md:25]:
```suggestion
| MLM    | Masked Language Modeling|
```

File: docs/source/resources.md [hunk 3]
```diff
@@ -14,6 +14,44 @@ Transformers4Rec is part of the NVIDIA Merlin ecosystem for Recommender Systems.
...
+| PLM    | Permutation Language Modelling|
```
Comment (gabrielspmoreira) [inline docs/source/resources.md:26]:
```suggestion
| PLM    | Permutation Language Modeling|
```

File: docs/source/resources.md [hunk 4]
```diff
@@ -14,6 +14,44 @@ Transformers4Rec is part of the NVIDIA Merlin ecosystem for Recommender Systems.
...
+| RTD    | Replacement Token Detection|
+
+
+In Transformers4Rec, we decouple the pre-training approaches from transformers architectures and provide `TransformerBlock` module that links the config class of the transformer architecture to the masking task. Transformers4Rec also defines a `transformer_registry` including pre-defined [`T4RecConfig`](https://nvidia-merlin.github.io/Transformers4Rec/main/api/transformers4rec.config.html#transformers4rec.config.transformer.T4RecConfig) constructors that automatically set the arguments of the related HuggingFace configuration classes.
```
Comment (gabrielspmoreira) [inline docs/source/resources.md:30]:
```suggestion
In Transformers4Rec, we decouple the pre-training approaches from transformers architectures and provide `TransformerBlock` module that links the config class of the transformer architecture to the masking task. Transformers4Rec also defines a `transformer_registry` including pre-defined [`T4RecConfig`](https://nvidia-merlin.github.io/Transformers4Rec/main/api/transformers4rec.config.html#transformers4rec.config.transformer.T4RecConfig) constructors that automatically set the arguments of the related HuggingFace Transformers' configuration classes. 
```

File: docs/source/resources.md [hunk 5]
```diff
@@ -14,6 +14,44 @@ Transformers4Rec is part of the NVIDIA Merlin ecosystem for Recommender Systems.
...
+|Accronym| Definition|
+|---------|--------------|
+| CLM    | Causal Language Modelling|
+| MLM    | Masked Language Modelling|
+| PLM    | Permutation Language Modelling|
+| RTD    | Replacement Token Detection|
+
+
+In Transformers4Rec, we decouple the pre-training approaches from transformers architectures and provide `TransformerBlock` module that links the config class of the transformer architecture to the masking task. Transformers4Rec also defines a `transformer_registry` including pre-defined [`T4RecConfig`](https://nvidia-merlin.github.io/Transformers4Rec/main/api/transformers4rec.config.html#transformers4rec.config.transformer.T4RecConfig) constructors that automatically set the arguments of the related HuggingFace configuration classes. 
+The table below represents the current supported architectures in Transformers4Rec and links them to the possible masking tasks. It also lists the pre-registered `T4RecConfig` classes in the colum `Registered`.
```
Comment (gabrielspmoreira) [inline docs/source/resources.md:31]:
```suggestion
The table below represents the current supported architectures in Transformers4Rec and links them to the possible masking tasks. It also lists the pre-registered `T4RecConfig` classes in the colum `Registered`.  
Tip: Registering HF Transformers config classes into Transformers4Rec is a good opportunity for your first contributions to the library ;)
```

File: docs/source/resources.md [hunk 6]
```diff
@@ -14,6 +14,44 @@ Transformers4Rec is part of the NVIDIA Merlin ecosystem for Recommender Systems.
...
+|  [Longformer](https://huggingface.co/transformers/model_doc/longformer.html#longformerconfig) |   ✅   | ✅     |   ❌   |  ❌  |   ✅   |
+| [MegatronBert](https://huggingface.co/transformers/model_doc/megatron_bert.html#megatronbertconfig) |   ❌   |  ✅    |   ❌   |  ✅  |   ❌   |
+|   [MPNet](https://huggingface.co/transformers/model_doc/mpnet.html#mpnetconfig)     |   ❌    |  ✅   |   ❌   |  ✅  |   ❌   |
+|   [RoBERTa](https://huggingface.co/transformers/model_doc/roberta.html#robertaconfig)   |   ❌    |  ✅   |   ❌   |  ✅  |   ✅   |
+|   [RoFormer](https://huggingface.co/transformers/model_doc/roformer.html#roformerconfig)  |   ✅    |  ✅   |   ❌   |  ✅  |   ❌   |
+| [Transformer-XL](https://huggingface.co/transformers/model_doc/transformerxl.html#transfoxlconfig)|   ✅    | ❌     |   ❌   |  ❌    |   ✅   |
+|   [XLNet](https://huggingface.co/transformers/model_doc/xlnet.html#xlnetconfig)    |   ✅    | ✅     |   ✅   |  ✅    |   ✅   |
+
+
+ **Note**: The following HF architectures will be supported in future release: `Reformer`, `Funnel Transformer`, `ELECTRA` and `ALBERT`.
```
Comment (gabrielspmoreira) [inline docs/source/resources.md:50]:
```suggestion
 **Note**: The following HF architectures will be registered in future release: `Reformer`, `Funnel Transformer`, `ELECTRA` and `ALBERT`. 
```

File: transformers4rec/config/transformer.py [hunk 7]
```diff
@@ -288,3 +288,105 @@ def build(
...
+        d_model,
+        n_head,
+        n_layer,
+        total_seq_length,
+        hidden_act="gelu",
+        initializer_range=0.01,
+        layer_norm_eps=0.03,
+        dropout=0.3,
+        pad_token=0,
+        log_attention_weights=True,
```
Comment (gabrielspmoreira) [inline transformers4rec/config/transformer.py:307]: Shouldn't `log_attention_weights` default be False?
I think it should be enabled only for debugging purposes

**Review by gabrielspmoreira — COMMENTED** [review]
Thanks @sararb  for the great job on mapping the compatible Transformer architectures architectures and training approaches. Just left some suggestions.

I particular, what is the issue to integrate ALBERT and ELECTRA?
If we can't include them in the library, them we are limited in paper reproducibility until they are added, right?

**Comment by sararb** [comment]
> Thanks @sararb for the great job on mapping the compatible Transformer architectures architectures and training approaches. Just left some suggestions.
> 
> I particular, what is the issue to integrate ALBERT and ELECTRA?
> If we can't include them in the library, them we are limited in paper reproducibility until they are added, right?

These two architectures are throwing an error when used with input_embeds, this was fixed in this [PR](https://github.com/huggingface/transformers/pull/13128) but was merged in the HF release `4.10.1`. So we'll need to upgrade the 4.9.* HF requirements to be able to use them

File: docs/source/resources.md [hunk 8]
```diff
@@ -14,6 +14,45 @@ Transformers4Rec is part of the NVIDIA Merlin ecosystem for Recommender Systems.
 - [Triton Inference Server](https://github.com/triton-inference-server/server). - Provides a cloud and edge inferencing solution optimized for both CPUs and GPUs. Transformers4Rec models can be exported and served with Triton.
 - [HugeCTR](https://github.com/NVIDIA/HugeCTR) - A GPU-accelerated recommender framework designed to distribute training across multiple GPUs and nodes and estimate Click-Through Rates (CTRs). 
 
+
+## Supported HuggingFace architectures and pre-training approaches
+
+Transformers4Rec supports the four following [masking tasks](https://nvidia-merlin.github.io/Transformers4Rec/main/model_definition.html#sequence-masking): 
+
+|Accronym| Definition|
```
Comment (benfred) [inline docs/source/resources.md:22]:
```suggestion
|Acronym| Definition|
```

**Review by benfred — APPROVED** [review]
This looks good to me!

**Pull request merged**